File: runinator-command-center/src/core/domain/models/pipeline/pipeline-trigger.ts
import type { JsonRecord } from "../../json";
import type { WorkflowTriggerKind } from "../workflow/trigger";

// a pipeline-level trigger. cron/manual start a run of this pipeline; a `chained` trigger is
// target-keyed (this pipeline starts when the configured source reaches a terminal state, carried in
// `configuration` as `source_workflow`/`source_pipeline` + `on`).
export interface PipelineTrigger {
  id: string | null;
  pipeline_id: string;
  kind: WorkflowTriggerKind;
  enabled: boolean;
  configuration: JsonRecord;
  next_execution: string | null;
  blackout_start: string | null;
  blackout_end: string | null;
  metadata: JsonRecord;
  created_at?: string | null;
  updated_at?: string | null;
}

File: runinator-models/src/pipelines/pipeline_trigger.rs
#[allow(unused_imports)]
use super::*;

/// a persisted pipeline-level trigger. mirrors [`crate::workflows::WorkflowTrigger`] but is owned by a
/// pipeline: cron/manual start a pipeline run for `pipeline_id`; a `chained` trigger is target-keyed
/// (`pipeline_id` is the pipeline to start) with its source and `on` selector in `configuration`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineTrigger {
    pub id: Option<Uuid>,
    pub pipeline_id: Uuid,
    pub kind: WorkflowTriggerKind,
    pub enabled: bool,
    #[serde(default)]
    pub configuration: Value,
    pub next_execution: Option<DateTime<Utc>>,
    pub blackout_start: Option<DateTime<Utc>>,
    pub blackout_end: Option<DateTime<Utc>>,
    #[serde(default)]
    pub metadata: Value,
    #[serde(default)]
    pub created_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub updated_at: Option<DateTime<Utc>>,
}

impl crate::validation::Validate for PipelineTrigger {
    fn validate(&self) -> Result<(), crate::validation::ValidationError> {
        crate::workflows::validate_trigger_window(self.blackout_start, self.blackout_end)?;
        crate::validation::dynamic_value("configuration", &self.configuration)?;
        crate::validation::dynamic_value("metadata", &self.metadata)?;
        Ok(())
    }
}
